README.md
Rotation.js <img src="http://benschwarz.github.io/bower-badges/badge@2x.png" width="130" height="30">
===========

Responsive and mobile enabled jQuery plugin to help create rotating content such as image carousel, slider or revolving testimonials.

## Main Features

* Quick to setup
* Highly configurable
* Semantic and minimal markup
* HTML5 Metadata support
* CSS3 transition animations with JavasScript fallback
* Mobile friendly with touch support and mouse fallback
* Accessibility support with ARIA markup
* Keyboard support
* Responsive support

## Usage

Include jQuery as the dependency

```javascript
  <script src="http://code.jquery.com/jquery-1.10.2.min.js"></script>
```

and then include `jquery.rotation.js` like so

```javascript
  <script src="jquery.rotation.js"></script>
```

and finally add basic stylesheet `jquery.rotation.css`

```
  <link rel="stylesheet" href="jquery.rotation.css">
```

### Markup
Then add semantic markup to describe your content, any item that requires rotation should be placed inside separate list element. For instance, to rotate quotes one can write markup such as

```html
<ul id="rotation" class="rotation-list">
  <li><blockquote><p>Morbi at odio .... </p></blockquote></li>
  <li><blockquote><p>Morbi at odio .... </p></blockquote></li>
  <li><blockquote><p>Morbi at odio .... </p></blockquote></li>
  ...
</ul>
```
Finally, initialize the **Rotation** plugin

```javascript
$("#rotation").rotation();
```

## Options

There are many options that you can specify for any **Rotation** instance. These options can be provided via constructor or API.

| Option              | Default    | Type   | Description
| -------             | ---------  | ------ | --------
| `autoRotate`        | `true`     | bool   | animate automatically
| `interval`          | `4000`     | int    | rotation interval (ms)
| `duration`          | `500`      | int    | transition speed (ms)
| `loop`              | `true`     | bool   | infinite scrolling
| `step`              | `1`        | int    | number of items to scroll
| `orientation`       | `horizontal` | string    | horizontal or vertical rotation
| `easing`            | `linear`   | string  | transition easing
| `containerClass`    | `rotation-container` | string | base container css class
| `keypress`          | `true`     | bool   | keyboard keypress navigation
| `touch`             | `true`     | bool   | enable touch events
| `touchMin`          | `30`       | int    | minimum amount of pixels to detect swipe gesture
| `touchMax`          | `320`      | int    | maximum amount of pixels to stop swipe gesture
| `touchDelay`        | `500`      | int    | swipe gesture maximum delay (ms)
| `responsive`        | `true`     | bool   | enable responsive support
| `responsiveDelay`   | `150`      | int    | resize event maximum delay (ms)
| `pagination`        | `true`     | bool   | show pagination
| `paginationNumbers` | `true`     | bool   | show pagination with numbers
| `paginationClass`   | `rotation-pagination` | string | top level css class for pagination
| `paginationItemClass` | `item` | string | css class for pagination item
| `paginationCurrentItemClass` | `current`  | string | css class for pagination current item
| `navControls`       | `true`    | bool   | show navigation controls
| `navControlsClass`  | `rotation-nav-controls` | string | top level css class for navigation controls
| `navControlsItemClass` | `item` | string | css class for navigation current item
| `navControlsNextClass` | `next` | string | css class for next nav control
| `navControlsPrevClass` | `prev` | string | css class for prev nav control
| `navControlsNextText`  | `>>`   | string | text for prev nav control
| `navControlsPrevText`  | `<<`   | string | text for prev nav control

Callback options

| Callback             | Default        | Type     | Description
| -------------------- | -------------- | -------- | --------
| `onInitStart`        | `function(){}` | function | before plugin initialization
| `onInitEnd`          | `function(){}` | function | after plugin initialization
| `onTransitionStart`  | `function(){}` | function | before rotation starts
| `onTransitionEnd`    | `function(){}` | function | after rotation ends
| `onSwipeLeft`        | `function(){}` | function | when swipe left gesture
| `onSwipeRight`       | `function(){}` | function | when swipe right gesture
| `onSwipeUp`          | `function(){}` | function | when swipe up gesture
| `onSwipeDown`        | `function(){}` | function | when swipe down gesture
| `onSwipe`            | `function(){}` | function | when any swipe gesture

### Initialization

The wide set of defaults provided by the **Rotation** can be easily extended by using object literal in the following ways:

```javascript
// create instance
var rotation = new Rotation($("#rotation"), {
  autoRotate: true
});

// or set the params on jquery instance
$("#rotation").rotation({
  autoRotate: true
});

// or modify global default params
Rotation.defeaults.autoRotate = true;
```

### API

## Metadata

| Attribute     | Description
| -------       | --------
| `data-init`   | initial params


## Events

There are number of custom events emitted by **Rotation** that you can listen for.

| Event                  | Description
| -----------------------| -------------------
| `rotation:swipe`       | when swipe gesture is performed
| `rotation:swipeleft`   | when swipe left
| `rotation:swiperight`  | when swipe right
| `rotation:swipeup`     | when swipe up
| `rotation:swipedown`   | when swipe down
| `rotation:scrollstart` | when mouse scroll is started
| `rotation:scrollstop`  | when mouse scroll is finished
| `rotation:transitionstart` | when animation is started
| `rotation:transitionend`   | when animation is finished
| `rotation:play`        | when auto rotation starts/resumes
| `rotation:pause`       | when auto rotation pauses
| `rotation:stop`        | when auto rotation is stopped

The `swipe`, `swipeleft`, `swiperight`, `swipeup`, `swipedown` events decorate the event object with `swipestart` and `swipeend` properties. Both `swipestart` and `swipeend` object literals contain following properties:

| Property     | Description
| -------      | --------
| `time`       | time when swipe gesture is registered
| `touch`      | touch object the event is registered with
| `coords`     | array with x and y coordinates for the touch event
| `origin`     | jquery object wrapper for the touch event

You can either register callback to listen for the custom event with `on` method:

```javascript
$("#rotation").rotation().
on("rotation:swipeleft", function (e) {
  console.log(e.swipestart.coords);
})
```

or by passing it to the constructor:

```javascript
$("#rotation").rotation({
  onSwipeLeft: function (e) { ... }
})
```

## Animation

The **Rotation** uses feature detection to turn CSS3 support in the following browsers:
`Chrome`, `Safari 4+`, `Firefox 4+`, `Android browsers`.

If you want to animate content inside each element use either after/beforeTransition callback
or alternatively css animations.

## Examples

### 1. Testimonials rotation:

### 2. Carousel

### 3. Slider

### 4. Twitter feed rotation


## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Copyright

Copyright (c) 2014 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
0.4.3 (Feb 2, 2014)

* Change to namespace custom events
* Add events unbinding to help setup plugin safely
* Change before/afterTransition callbacks to onTransitionStart, onTransitionEnd
* Add rotation:transitionstart, rotation:transition, rotation:transitionend events
* Add logger

0.4.2 (Jan 14, 2014)

* Fix pause behaviour
* Add CSS3 features detection
* Add ability to specify orientation for the slide animation
* Add easing option
* Add all callbacks from options

0.4.1 (Jan 03, 2014)

* Add options validation
* Change markup to include viewport and new navigation controls
* Add accessibility
* Add ability to scroll items
* Add responsive support

0.4.0 (Dec 30, 2013)

* Add touch support


test/init.js
(function () {
  var instance;

  module("initialization", {
    setup: function () {
      instance = new Rotation();
    }
  });

  test("constructor", function () {
    var element = $(".rotation-container"),
    rotation = new Rotation(element);
    equal(rotation.itemsCount(), 2, "equal items");
  });

  test("current", function () {
    var element = $(".rotation-container"),
    rotation = new Rotation(element);
    equal(rotation.current(), 0, "0 index");
  });
})();


Gruntfile.js
module.exports = function (grunt) {
  'use strict';

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    uglify: {
      options: {
        banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' + 
          '<%= grunt.template.today("yyyy-mm-dd") %> */'
      },
      build: {
        files: {
          'jquery.rotation.min.js' : ['jquery.rotation.js']
        }
      }
    },
    jshint: {
      build: {
        src: ["jquery.rotation.js", "Gruntfile.js", "test/*.js"],
        jshintrc: true
      }
    },
    regarde: {
      uglify: {
        files: ['jquery.rotation.js'],
        tasks: ['uglify', 'jshint']
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-regarde');

  grunt.registerTask('dev', ['regarde', 'jshint']);

  grunt.registerTask('default', ['uglify']);
};


jquery.rotation.js
/*
 * rotation.js 0.4.3
 *
 * (c) 2014 Piotr Murach
 *
 * Creates rotation of elements with various transition events.
 */
;(function ($, window, document, undefined) {
  'use strict';

  var
    name = 'rotation',
    defaults = {
    // String: default namespace used for classes and events
    namespace: 'rotation',
    // animate automatically
    autoRotate: true,
    // rotation interval (ms)
    interval: 4000,
    // transition speed (ms)
    duration: 300,
    // transition easing
    easing: 'linear',
    transitionIn: 'fadeIn',
    transitionOut: 'fadeOut',
    // horizontal or vertical rotation
    orientation: 'horizontal',
    // auto rotation pauses on hover
    pauseOnHover: true,
    // auto rotation stops on hover
    stopOnHover: false,
    // mouse scroll content
    scrolling: true,
    // enable keyboard events
    keypress: true,
    // enable touch events
    touch: true,
    // it isn't a swipe over this time
    touchDelay: 500,
    // minimum amount of pixels for swipe gesture
    touchMin: 30,
    // maximum amount of pixels for swipe gesture
    touchMax: 320,
    // circular rotation
    loop: true,
    visibleItems: 1,
    step: 1,
    itemWidth: 200,
    // eanble responsive support
    responsive: true,
    // resize delay (ms)
    responsiveDelay: 150,
    // main container css
    containerClass: 'rotation-container',

    // display controls
    pausePlayControl: true,
    pausePlayControlClass: 'rotation-play-pause',
    pauseText: 'Pause',
    playText: 'Play',
    // navigation control arrows
    navControls          : true,
    navControlsClass     : 'rotation-nav-controls',
    navControlsItemClass : 'rotation-item',
    navControlsNextClass : 'rotation-next',
    navControlsPrevClass : 'rotation-prev',
    navControlsNextText  : '>>',
    navControlsPrevText  : '<<',
    // pagination controls
    pagination: true,
    paginationNumbers: true,
    paginationClass: 'rotation-pagination',
    paginationItemClass: 'rotation-page',
    paginationCurrentItemClass: 'current',

    // callbacks
    onInitStart: $.noop,
    onInitEnd: $.noop,
    onTransitionStart: $.noop,
    onTransitionEnd: $.noop,
    onPlay: $.noop,
    onPause: $.noop,
    onSwipeLeft: $.noop,
    onSwipeRight: $.noop,
    onSwipeUp: $.noop,
    onSwipeDown: $.noop,
    onSwipe: $.noop
  };

  var Rotation = function (container, options) {
    var
      self = this, markup, viewport;

    this.options = $.extend({}, defaults, options);

    markup   = $('<section/>', {'class': this.getOption("containerClass")});
    viewport = $('<div/>', {'class': 'rotation-viewport'});

    this._defaults       = defaults;
    this.namespace       = this.options.namespace;
    this.currentIndex    = 0;
    this.isAnimating     = false; // control animation
    this.touchSupported  = 'ontouchend' in document;
    this.itemsContainer  = container;
    this.$itemsContainer = $(this.itemsContainer);
    this.$itemsViewport  = this.$itemsContainer.wrap(viewport).parent();
    this.$items          = this.$itemsContainer.children();
    this.$container      = this.$itemsViewport.wrap(markup).parent();
    this.metadata        = this.$container.data("rotation-options");
    this.cssTransition   = this.hasCSS('transition');
    this.cssTransform    = this.hasCSS('transform');
    this.cssSupported    = !!(this.cssTransition && this.cssTransform);
    this.cssTransEvent   = this.transitionEndEvent();

    // remove whitespace
    this.$items.detach();
    this.$itemsContainer.empty();
    this.$itemsContainer.append(this.$items);

    this.validate();

    this.triggerCustomEvent(this.$itemsContainer, $.Event(self.namespace + ':initstart'));
    this.options.onInitStart.call(this);

    this.init();
    this.build();
    this.bindEvents();
    this.play();

    this.triggerCustomEvent(this.$itemsContainer, $.Event(this.namespace + ':initend'));
    this.options.onInitEnd.call(this);

    // public API
    return {
      // appends item dynamically to container
      add: function (item, callback) {

      },

      play: function () {
        self.play();
      },

      pause: function () {
        self.pause();
      },

      itemsCount: function () {
        return self.itemsCount();
      },

      current: function () {
        return self.currentIndex;
      },

      setCurrent: function (idx, callback) {
        self.currentIndex = idx;
        // self.rotateTo(idx)
        if (callback) { callback.call(); }
      },

      next: function () {
        self.rotate(+self.getOption("step"));
      },

      previous: function () {
        self.rotate(-self.getOption("step"));
      },

      destroy: function () {
        //self.destroy()
      }
    };
  };

  Rotation.VERSION = '0.4.3';

  $.extend(Rotation.prototype, {

    getOption: function (option) {
      return this.options[option];
    },

    setOption: function (option, value) {
      this.options[option] = value;
      return value;
    },

    triggerCustomEvent: function (obj, event) {
      obj.trigger(event);
    },

    itemsCount: function () {
      return this.$items.length;
    },

    getItemByIndex: function (index) {
      return this.$items.eq(index);
    },

    lockAnimation: function () {
      this.isAnimating = true;
    },

    isLocked: function () {
      return !!this.isAnimating;
    },

    unlockAnimation: function () {
      this.isAnimating = false;
    },

    /*
    * Test for css property against vendor specific ones and non-vendor one.
    */
    hasCSS: function (prop) {
      var
        el       = document.createElement('rotation'),
        prefixes = ['Khtml', 'Webkit', 'ms', 'Moz', 'O'],
        lookup   = [''].concat(prefixes).concat(['']),
        capProp  = prop.charAt(0).toUpperCase() + prop.slice(1),
        props    = (prop + ' ' + prefixes.join(capProp + ' ') + capProp).split(' '),
        prefix, i;

      $(document.body).prepend(el);

      for (i in props) {
        prefix = lookup[i].toLowerCase();
        if (prefix !== '') { prefix = '-' + prefix + '-'; }

        if (props[i] in el.style) {
          return prefix + prop;
        }
      }

      $(document.body).children()[0].remove();

      return false;
    },

    /*
    * Test for css transtion end event.
    */
    transitionEndEvent: function () {
      var
        el = document.createElement('rotation'),
        transEndEventNames = {
          'WebkitTransition' : 'webkitTransitionEnd',
          'MozTransition'    : 'transitionend',
          'OTransition'      : 'oTransitionEnd',
          'msTransition'     : 'MSTransitionEnd',
          'transition'       : 'transitionend'
        };

      for (var name in transEndEventNames) {
        if (el.style[name] !== undefined) {
          return transEndEventNames[name];
        }
      }

      $(document.body).children()[0].remove();

      return false;
    },

    /*
     * Initialize
     */
    init: function () {
      var self = this;

      this.$window       = $(window);
      this.windowHeight  = this.$window.height();
      this.windowWidth   = this.$window.width();
      this.viewportWidth = this.$itemsViewport.outerWidth(true);
      this.viewportHeight = this.$itemsViewport.outerHeight(true);
      this.itemWidth     = this.viewportWidth / this.getOption("visibleItems");
      this.itemHeight    = this.viewportHeight / this.getOption("visibleItems");
      this.itemsWidth    = this.itemWidth * this.itemsCount();
      this.itemsHeight   = this.itemHeight * this.itemsCount();

      this.initIdsOnElements();

      this.$itemsContainer.width(this.itemsWidth);
      this.$items.width(this.itemWidth);
      this.$items.hide();
      var itemsToShow = this.$items.slice(0, this.getOption("visibleItems"));

      $.each(itemsToShow, function (idx, item) {
        $(item).css("left", idx * self.itemWidth).show();
      });
    },

    initIdsOnElements: function () {
      var i;

      for (i = 0; i < this.itemsCount(); i++) {
        this.getItemByIndex(i).addClass('rotation-item-' + (i + 1));
      }
    },

    /*
     * Validate options
     */
    validate: function () {
      var
        responsive  = this.getOption("respnosive"),
        orientation = this.getOption("orientation");

      if (typeof responsive !== 'boolean') {
        this.setOption("responsive", Boolean(responsive));
      }

      if (orientation !== 'horizontal' && orientation !== 'vertical') {
        this.setOption("orientation", 'horizontal');
      }
    },

    /*
     * Bind plugin event listeners
     */
    bindEvents: function () {

      if (this.getOption("responsive")) {
        this.bindResizeEvent();
      }

      if (this.getOption("pauseOnHover")) {
        this.bindMouseEvents();
      }

      if (this.getOption("scrolling")) {
        this.bindScrolling();
      }

      if (this.getOption("keypress")) {
        this.bindKeyboardEvents();
      }

      if (this.getOption("touch")) {
        this.bindTouchEvents();
      }
    },

    /*
     * Unbind already defiend plugin event listeners
     */
    unbindEvents: function () {
      $(document).off('keydown.' + this.namespace);
      this.$itemsContainer.off('swipeleft swiperight swipeup swipedown .' + this.namespace);
      this.$navContainer.find('a').off('click touchstart .' + this.namespace);
      this.$paginationContainer.find('a').off('click touchstart .' + this.namespace);
    },

    bindResizeEvent: function () {
      var self = this,
          timer;

      self.$window.on("resize." + this.namespace, function (e) {
        if (e.originalEvent) {
          var width = self.$window.width(),
              height = self.$window.height();

          if (width !== self.windowWidth || height !== self.windowHeight) {
            self.windowWidth = width;
            self.windowHeight = height;
            clearTimeout(timer);
            timer = setTimeout(function () {
              self.init();
              self.rotate(0);
            }, self.getOption("responsiveDelay"));
          }
        }
      });
    },

    bindKeyboardEvents: function () {
      var self = this;

      $(document).on('keydown.' + this.namespace, function (e) {
        // next or up
        if (e.keyCode === 39 || e.keyCode === 38) {
          self.rotate(+self.getOption("step"));
        }
        // prev or down
        if (e.keyCode === 37 || e.keyCode === 40) {
          self.rotate(-self.getOption("step"));
        }
      });
    },

    bindMouseEvents: function () {
      var self = this;

      this.$itemsContainer.
        on("mouseenter." + this.namespace, function () {
          self.setOption("pause", true);
        }).
        on("mouseleave." + this.namespace, function () {
          self.setOption("pause", false);
        });
    },

    bindScrolling: function () {
      var self = this;

      this.$itemsContainer.on("mousewheel." + this.namespace, function (e) {
        e.preventDefault();
        self.rotate(self.currentIndex + self.getOption("step"));
      });
    },

    /*
     * Move elemnts on swipe.
     */
    bindTouchEvents: function () {
      var
        self = this,
        elapsedTime,
        touchStart,
        touchEnd,
        touchDistance,
        touchDeltaX,
        touchDeltaY,
        touchDelay = self.getOption("touchDelay"),
        touchMin   = self.getOption("touchMin"),
        touchMax   = self.getOption("touchMax"),
        namespace  = self.getOption("namespace"),
        touchStartEvent = self.touchSupported ? 'touchstart' : 'mousedown',
        touchStopEvent  = self.touchSupported ? 'touchend' : 'mouseout',
        touchMoveEvent  = self.touchSupported ? 'touchmove' : 'mousemove',

        chooseTouchEvent = function (e) {
          if (e.pageX) { return e; }
          return e.originalEvent.touches[0] || e.originalEvent.changedTouches[0];
        },

        createTouch = function (e) {
          var touch = chooseTouchEvent(e);
          return {
            time:   new Date().getTime(),
            touch:  touch,
            coords: [touch.pageX, touch.pageY],
            origin: $(e.target)
          };
        },

        onTouchStart = function (e)  {
          e.preventDefault();

          touchStart = createTouch(e);
        },

        onTouchMove = function (e) {
          if (!touchStart) {
            return;
          }

          touchEnd = createTouch(e);

          // prevent scrolling
          if (Math.abs(touchStart.coords[0] - touchEnd.coords[0]) > 10) {
            e.preventDefault();
          }
        },

        onTouchEnd = function (e) {
          e.preventDefault();

          if (!touchStart || !touchEnd) {
            return;
          }

          elapsedTime   = touchEnd.time - touchStart.time;
          touchDeltaX   = Math.abs(touchStart.coords[0] - touchEnd.coords[0]);
          touchDeltaY   = Math.abs(touchStart.coords[1] - touchEnd.coords[1]);
          touchDistance = Math.sqrt(touchDeltaX * touchDeltaX + touchDeltaY * touchDeltaY);

          // moved less than 15 pixels and touch duration is less than 100ms
          // trigger tap event on the element
          if (touchDistance < 15 && elapsedTime < 100)  {
            touchStart.origin.trigger('click');
            return;
          }

          if (elapsedTime < touchDelay &&
              touchDistance >= touchMin &&
              touchDistance <= touchMax) {

            var events = [namespace + ':swipe'];

            if (touchDeltaX >= touchMin && touchDeltaY < touchMin) {
              self.setOption("pause", false);

              if (touchStart.coords[0] > touchEnd.coords[0]) {
                events.push(namespace + ":swipeleft");
                self.rotate(-self.getOption("step"));
                self.getOption('onSwipeLeft').call(self);
              } else {
                events.push(namespace + ":swiperight");
                self.rotate(+self.getOption("step"));
                self.getOption('onSwipeRight').call(self);
              }
            } else if(touchDeltaY >= touchMin && touchDeltaX < touchMin) {
              self.setOption("pause", false);

              if (touchStart.coords[1] < touchEnd.coords[1]) {
                events.push(namespace + ":swipedown");
                self.rotate(-self.getOption("step"));
                self.getOption('onSwipeDown').call(self);
              } else {
                events.push(namespace + ":swipeup");
                self.rotate(+self.getOption("step"));
                self.getOption('onSwipeUp').call(self);
              }
            }

            $.each(events, function (index, eventName) {
              var eventObject = $.Event(eventName, {
                target: e.target,
                swipestart: touchStart,
                swipeend: touchEnd,
              });
              self.triggerCustomEvent(self.$itemsContainer, eventObject);
            });
          }
        };

      self.$itemsContainer
        .on(touchStartEvent + '.' + this.namespace, onTouchStart)
        .on(touchMoveEvent  + '.' + this.namespace, onTouchMove)
        .on(touchStopEvent  + '.' + this.namespace, onTouchEnd);
    },

    /*
     * Main content rotation.
     */
    rotate: function (direction, callback) {
      // nothing to rotate
      if (this.itemsCount() <= 1) { return; }

      // Pause if user has interacted
      if (this.getOption("pause")) { return; }

      if (direction) { this.pause(); }


      var
        namespace       = this.getOption("namespace"),
        currentIndex    = this.currentIndex,
        itemsCount      = this.itemsCount(),
        previousElement = this.getItemByIndex(currentIndex),
        dimension, element, distance;

      dimension = this.getOption("orientation") === 'horizontal' ? this.itemWidth : this.itemHeight;

      if (currentIndex === 0 && direction < 0) {
        currentIndex = itemsCount - 1;
      } else {
        currentIndex = (direction ? currentIndex + direction : ++currentIndex) % itemsCount;
      }

      element = this.getItemByIndex(currentIndex);
      distance = direction ? direction * dimension : dimension;

      this.$itemsContainer.stop();

      if (this.isLocked()) { return; }

      this.lockAnimation();

      this.getOption("onTransitionStart").call(this);
      this.triggerCustomEvent(this.$itemsContainer, $.Event(namespace + ':transitionstart'));

      this.animations.slide.call(this, previousElement, element, distance, currentIndex);

      if (this.getOption("pagination")) {
        this.setPaginationCurrentItem(direction);
      }

      // callbacks
      if (callback && (typeof callback === 'function')) { callback(); }

      if (direction) { this.play(); }
    },

    animations: {
      slide: function(previousElement, element, distance, currentIndex) {
        var
          self = this,
          duration    = this.getOption("duration"),
          easing      = this.getOption("easing"),
          orientation = this.getOption("orientation"),
          namespace   = this.getOption("namespace"),
          promises    = [];

        if (orientation === 'horizontal') {
          promises.push(
            $(previousElement).css({left: 0})
            .animate({left: -distance}, {duration: duration, easing: easing}).promise()
          );

          promises.push(
            $(element).css({left: distance}).show()
            .animate({left: 0},{duration: duration, easing: easing}).promise()
          );
        } else {
          promises.push(
            $(previousElement).css({top: 0})
            .animate({top: -distance}, {duration: duration}).promise()
          );

          promises.push(
            $(element).css({top: distance}).show()
            .animate({top: 0},{duration: duration}).promise()
          );
        }

        $.when.apply(null, promises).done(function () {
          self.currentIndex = currentIndex;
          self.unlockAnimation();
          self.getOption("onTransitionEnd").call(self);
          self.triggerCustomEvent(self.$itemsContainer, $.Event(namespace + ':transitionend'));
        });
      }
    },

    play: function () {
      var self = this,
          namespace = self.getOption("namespace");

      if (this.getOption("autoRotate")) {
        this.auto = setInterval(function () {
          self.rotate();
        }, this.getOption("interval"));
        this.getOption("onPlay").call(this);
        self.triggerCustomEvent(self.$itemsContainer, $.Event(namespace + ':play'));
      }
    },

    pause: function () {
      var self = this,
          namespace = self.getOption("namespace");

      if (self.auto) {
        self.auto = clearInterval(self.auto);
      }
      this.getOption("onPause").call(this);
      self.triggerCustomEvent(self.$itemsContainer, $.Event(namespace + ':pause'));
    },

    setPaginationCurrentItem: function (direction) {
      var
        currentClass        = this.getOption("paginationCurrentItemClass"),
        currentItem         = $("a." + currentClass, this.$paginationContainer).parent('li'),
        nextIndex, nextItem, nextPage, selector;

      selector = direction < 0 ? "last" : "first";
      if (direction) {
        nextIndex = direction < 0 ? this.currentIndex + direction : this.currentIndex + direction - 1;
      } else {
        nextIndex = this.currentIndex;
      }
      nextItem = currentItem.siblings().eq(nextIndex);
      nextPage = nextItem.length ? nextItem.find('a') : $("a:" + selector, this.$paginationContainer);

      $('a.' + currentClass, this.$paginationContainer).removeClass(currentClass);
      nextPage.addClass(currentClass);
    },

    /*
     * Build UI in DOM
     */
    build: function () {
      if (this.getOption("navControls")) {
        this.buildNavControls();
      }

      if (this.getOption("pagination")) {
        this.buildPagination();
      }
    },

    /*
     * Builds rotator navigation.
     */
    buildNavControls: function () {
      if (this.itemsCount() <= 1) { return; }

      var self = this,
          container = this.$container,
          startEvents = 'click touchstart .' + this.namespace,
          item;

      self.$navContainer = $('<ul/>', {'class': this.getOption("navControlsClass")});
      $('<nav/>').wrapInner(self.$navContainer).appendTo(container);

      item = $('<a/>', {
        'href': '#',
        'class': this.getOption("navControlsItemClass") + ' ' + this.getOption("navControlsPrevClass"),
        'data-direction': -this.getOption("step"),
        'html': this.getOption("navControlsPrevText")
      });
      $('<li/>').wrapInner(item).appendTo(self.$navContainer);
      item = $('<a/>', {
        'href': '#',
        'class': this.getOption("navControlsItemClass") + ' ' + this.getOption("navControlsNextClass"),
        'data-direction': +this.getOption("step"),
        'html': this.getOption("navControlsNextText")
      });
      $('<li/>').wrapInner(item).appendTo(self.$navContainer);


      self.$navContainer.children().find('a').on(startEvents, function (e) {
        e.preventDefault();
        var direction = +$(this).data('direction');

        self.rotate(direction);
      });
    },

    /*
     * Builds rotator pagination.
     */
    buildPagination: function () {

      if (this.itemsCount() <= 1) { return; }

      var
        self = this,
        container = this.$container,
        startEvents = 'click touchstart .' + this.namespace,
        item, link, i, navItems;

      self.$paginationContainer = $('<ol/>', {
        'class': this.getOption("paginationClass"),
        'role': 'navigation',
        'aria-labelledby': 'paginglabel'
      });
      self.$paginationContainer.appendTo(container);

      for (i = 0; i < this.itemsCount(); i++) {
        item = $('<li/>');
        link = $('<a/>', {
          'href': '#rotation-item-' + i,
          'data-index': i,
          'text': this.getOption("paginationNumbers") ? i+1 : '',
          'class': this.getOption("paginationItemClass")
        });
        link.appendTo(item);
        item.appendTo(self.$paginationContainer);
      }

      navItems = $('li', self.$paginationContainer).children();
      navItems.eq(0).addClass(this.getOption("paginationCurrentItemClass"));

      navItems.on(startEvents, function (e) {
        e.preventDefault();
        var
          currentElement = $(this),
          currentClass   = self.getOption("paginationCurrentItemClass"),
          pageIndex      = +currentElement.data('index'),
          direction, currentIndex;

        currentIndex = +currentElement.
          parent('li').siblings('li').find('a.'+currentClass).data('index');
        direction =  pageIndex - currentIndex;

        self.rotate(direction);
      });
    }

  });

  /*
   * logger
   */
  function log() {
    if (window.console && console.log) {
      console.log(name + ' : ' + Array.prototype.join.call(arguments, ' '));
    }
  };

  Rotation.defaults = Rotation.prototype.defaults;

  $.fn[name] = function (options) {
    return this.each(function () {
      var container = this,
          instance = $.data(container, name);

      if (!instance) {
        $.data(container, name, new Rotation(container, options));
      }
    });
  };

  window.Rotation = Rotation;

})(jQuery, window, document);
